database/engine/rrdengineapi.c

```diff
@@ -1034,3 +1034,264 @@ void rrdeng_prepare_exit(struct rrdengine_instance *ctx)
+    pg_cache_insert(ctx, page_index, descr);
+    // Try to update the time start and end of the metric.
+    pg_cache_add_new_metric_time(page_index, descr);
+
+    debug(D_REPLICATION, "%s Commiting page in dbengine....", REPLICATION_MSG);
+    if (likely(descr->page_length)) {
+        int page_is_empty;
+
+        rrd_stat_atomic_add(&ctx->stats.metric_API_producers, -1);
+
+        page_is_empty = page_has_only_empty_metrics(descr);
+        if (page_is_empty) {
+            debug(D_REPLICATION, "%s Past metrics page has empty metrics only, deleting:", REPLICATION_MSG);
+            pg_cache_put(ctx, descr);
+            pg_cache_punch_hole(ctx, descr, 1, 0, NULL);
+        } else {
+            rrdeng_commit_page(ctx, descr, dim_past_data->page_correlation_id);
+        }
+    } else {
+        freez(descr->pg_cache_descr->page);
+        rrdeng_destroy_pg_cache_descr(ctx, descr->pg_cache_descr);
+        freez(descr);
+    }
+    debug(D_REPLICATION, "%s Page Commited -  Dimension \"%s\".\"%s\" metrics page commited in memory.", REPLICATION_MSG, rd->rrdset->id, rd->id);
+    // TBR: Only for debug
+    debug(D_REPLICATION,"%s REP: OBSERVE dimension (%s.%s) in time_interval[%ld, %ld] #samples(%lu)....END", REPLICATION_MSG, rd->rrdset->id, rd->id, (time_t)(descr->start_time/USEC_PER_SEC), (time_t)(descr->end_time/USEC_PER_SEC), (descr->page_length/sizeof(storage_number)));
+}
+
+void rrdeng_store_past_metrics_page_finalize(RRDDIM_PAST_DATA *dim_past_data, REPLICATION_STATE *rep_state){
+    UNUSED(rep_state);
+    struct pg_cache_page_index* page_index;
+    RRDDIM *rd = dim_past_data->rd;
+    page_index = dim_past_data->rd->state->page_index;
+#ifndef NETDATA_INTERNAL_CHECKS
+    UNUSED(rd);
+#endif
+
+    uv_rwlock_wrlock(&page_index->lock);
+    --page_index->writers;
+    uv_rwlock_wrunlock(&page_index->lock);
+    debug(D_REPLICATION, "%s Finalize operation -  Dimension \"%s\".\"%s\" metrics page completed.", REPLICATION_MSG, rd->rrdset->id, rd->id);
+}
+
+//It updates the dim_past_data according to new start and end times
+void modify_dim_past_data(RRDDIM_PAST_DATA *dim_past_data, usec_t start_time, usec_t end_time)
+{
+    uint64_t start, end, new_start, new_end, new_entries;
+    start = dim_past_data->start_time / USEC_PER_SEC; //gap time start
+    end = dim_past_data->end_time / USEC_PER_SEC;     //gap time end
+    new_start = start_time / USEC_PER_SEC;
+    new_end = end_time / USEC_PER_SEC;
+    new_entries = (uint64_t)(new_end - new_start) / dim_past_data->rd->update_every + 1;
+#ifndef NETDATA_INTERNAL_CHECKS
+    UNUSED(end);
+#endif
+
+    dim_past_data->page_length = new_entries * sizeof(storage_number);
+    if(new_start != start) {
+        storage_number *sn = dim_past_data->page;
+        dim_past_data->page = &sn[((new_start - start) / dim_past_data->rd->update_every) * sizeof(storage_number)];
+    }
+    dim_past_data->start_time = new_start * USEC_PER_SEC;
+    dim_past_data->end_time = new_end * USEC_PER_SEC;
```

Comment:
In lines `1183` and `1184` you divided the arguments `start_time` and `end_time` by `USEC_PER_SEC` to set two variables. Considering that you did not change their values, I suggest to use them directly instead to multiply `new_start` and `new_end` to retrieve values that are still available.